sql/src/main/java/io/druid/sql/calcite/schema/DruidSchema.java

```diff
@@ -59,42 +58,66 @@
 import io.druid.sql.calcite.view.DruidViewMacro;
 import io.druid.sql.calcite.view.ViewManager;
 import io.druid.timeline.DataSegment;
-import org.apache.calcite.schema.Function;
 import org.apache.calcite.schema.Table;
 import org.apache.calcite.schema.impl.AbstractSchema;
 import org.joda.time.DateTime;
 
+import java.io.IOException;
+import java.util.Comparator;
 import java.util.EnumSet;
-import java.util.List;
+import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutorService;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
 
 @ManageLifecycle
 public class DruidSchema extends AbstractSchema
 {
+  // Newest segments first, so they override older ones.
```

Comment:
I fail to see where exactly this order is used and why inner maps in `segmentSignatures` need to be TreeMaps, @gianm could you please explain?

Reply:
The idea is that Druid allows different segments to have different column types, and even allows the same column to have different types across different segments. But for SQL planning and metadata reporting purposes (e.g. INFORMATION_SCHEMA.COLUMNS) we can only report one type per column. We are preferring later segments (based on segment interval) over earlier ones if there's such a conflict, and achieve that by having `SEGMENT_ORDER` be an order that puts the newest segments first, and using it to affect the iteration order of `segmentMap.values()` in `buildDruidTable`.

Reply:
There are more uses of `TreeMap` and `TreeSet` in this class. Including some that are constructed without `SEGMENT_ORDER`. Maybe some of them don't need to be sorted collections.

Reply:
It also might be that based on how often `buildDruidTable()` is called compared to some other methods in `DruidSchema` that it's better to use HashMap for segmentMetadataInfo's values, and perform more work in `buildDruidTable()`, like having `Map<String, Pair<DataSegment, ValueType>> columnTypes`, iterating over `segmentMap.entrySet()` and use `columnTypes.compute()` with replacement decided comparing `DataSegment` objects.

Reply:
The latter may not improve performance, but be necessary for #6827, because `segmentMetadataInfo` may need to become `ConcurrentHashMap<String, ConcurrentHashMap<DataSegment, SegmentMetadataHolder>>` to enable safe iteration concurrent with modification. Assuming that we want to avoid `ConcurrentSkipListMap` because it's not memory efficient and slow.